lib/src/rules/avoid_renaming_method_parameters.dart

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/analyzer.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r"Don't rename parameters of overridden methods.";
+
+const _details = r'''**DON'T** rename parameters of overridden methods.
+
+Methods that override another method, but do not have their own documentation
+comment, will inherit the overridden method's comment when dartdoc produces
+documentation. If the inherited method contains the name of the parameter (in
+square brackets), then dartdoc cannot link it correctly.
+
+**BAD:**
+```
+abstract class A {
+  m(a);
+}
+
+abstract class B extends A {
+  m(b);
+}
+```
+
+**GOOD:**
+```
+abstract class A {
+  m(a);
+}
+
+abstract class B extends A {
+  m(a);
+}
+```
+
+''';
+
+class AvoidRenamingMethodParameters extends LintRule {
+  AvoidRenamingMethodParameters()
+      : super(
+            name: 'avoid_renaming_method_parameters',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  visitMethodDeclaration(MethodDeclaration node) {
+    if (node.isStatic) return;
+    if (node.documentationComment != null) return;
+
+    ClassDeclaration clazz = node.parent;
+
+    if (clazz.element.isPrivate) return;
+
+    final parentMethod = clazz.element
+        .lookUpInheritedMethod(node.name.name, clazz.element.library);
+
+    if (parentMethod == null) return;
+
+    final parameters = node.parameters.parameters
+        .where((p) => p.kind != ParameterKind.NAMED)
+        .toList();
+    final parentParameters = parentMethod.parameters
+        .where((p) => p.parameterKind != ParameterKind.NAMED)
+        .toList();
+    for (var i = 0; i < parameters.length; i++) {
```

Comment:
A minor point, but it would be slightly more efficient to compute the number of parameters that need to be matched up outside the loop rather than checking every time through the loop:
```
int count = math.min(parameters.length, parentParameters.length);
for (var i = 0; i < count; i++) {
```
Also, I have a small concern that this could be noisy in the face of errors. If the method from the subclass is not a valid override of the method from the superclass, then the user will already have errors reported, and having this lint produce more diagnostics would not be helpful.

The only solution I know of to that is to check whether the override is valid, but analyzer doesn't have an API for that and I don't want to duplicate that code here. I think the best thing we can do short term is to add a TODO comment to capture this concern so that we remember to address it later.

Reply:
> noisy in the face of errors

Lints are usually sorted below warnings. Interesting to ponder, but I don't feel this is a big deal...

Reply:
Generally I agree, though we do get a fair bit of feedback from users when we produce multiple diagnostics for a single problem, so it's worth taking it into account. But that's part of why I only asked for a TODO comment.